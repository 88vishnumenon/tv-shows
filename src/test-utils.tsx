import { Provider } from "react-redux";
import { RootState, getStoreWithState, store } from "./store/store";
import { BrowserRouter as Router } from "react-router-dom";
import { render } from "@testing-library/react";
import * as React from "react";
import { TVShow } from "./types/types";

export function renderWithContext(
  element: React.ReactElement,
  state?: RootState
) {
  const store = getStoreWithState(state);
  const utils = render(
    <Provider store={store}>
      <Router>{element}</Router>
    </Provider>
  );
  return { store, ...utils };
}

export function getStateWithItems(
  tvShows: TVShow[]
): RootState {
  const state: RootState = {
    tvshow: {
      showList: tvShows,
      generesList: ["DRAMA"],
      searchString: "",
      error: false,
    },
  };

  return state;
}
