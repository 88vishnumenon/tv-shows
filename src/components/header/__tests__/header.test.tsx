import userEvent from "@testing-library/user-event";
import { renderWithContext } from "../../../test-utils";
import Header from "../header";
import { cleanup, screen, waitFor } from "@testing-library/react";
import * as React from "react";

let mockPathName = "/testroute";
const mockUsedNavigate = jest.fn();

afterEach(() => {
  jest.clearAllMocks();
  cleanup();
});

jest.mock("react-router", () => ({
  ...(jest.requireActual("react-router") as {}),
  useLocation: jest.fn().mockImplementation(() => {
    return { pathname: mockPathName };
  }),
  useNavigate: () => mockUsedNavigate,
}));

describe("render header component", () => {
  it("snap shot testing", () => {
    const { asFragment } = renderWithContext(<Header></Header>);
    expect(asFragment()).toMatchSnapshot();
  });
  it("should not render homebutton in homepage", () => {
    renderWithContext(<Header></Header>);
    const headerBtn = screen.queryByTestId("header-home-btn");
    expect(headerBtn).toBe(null);
  });
  it("should render home buuton in details page ", () => {
    mockPathName = "/showDetails";
    renderWithContext(<Header></Header>);
    const headerBtn = screen.queryByTestId("header-home-btn");
    expect(headerBtn).toBeInTheDocument();
  });

  it("should direct to home page on home button click", async () => {
    mockPathName = "/showDetails";
    renderWithContext(<Header></Header>);
    const headerBtn = screen.getByTestId("header-home-btn");
    userEvent.click(headerBtn);
    await waitFor(() => expect(mockUsedNavigate).toHaveBeenCalled(), {
      timeout: 500,
    });
  });
});
