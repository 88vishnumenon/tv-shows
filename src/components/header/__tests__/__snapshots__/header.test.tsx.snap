// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render header component snap shot testing 1`] = `
<DocumentFragment>
  <section
    class="header"
  />
</DocumentFragment>
`;
